### PR TITLE
Fixes for HTLC db issues

### PR DESCRIPTION
**Pull request by rustyrussell**
Puzzled about two PR's, I wrote a stress test, which revealed a number of bugs, including those which caused the horrific issues #1960 and #1993.

It also triggers a yet-to-be-diagnosed occasional issue where we dislike the peer's option_dataloss_protect; I have temporarily disabled that check in order to resolve these more important issues, but am now chasing that one.

Fixes: #1960 
Fixes: #1993

### CODE REVIEW

File: wallet/wallet.c [hunk 1]
```diff
@@ -1378,19 +1385,20 @@ static void fixup_hin(struct wallet *wallet, struct htlc_in *hin)
 	if (hin->failcode || hin->failuremsg)
 		return;
 
-	hin->failcode = WIRE_TEMPORARY_CHANNEL_FAILURE;
+	hin->failcode = WIRE_TEMPORARY_NODE_FAILURE;
```
Comment (cdecker) [inline wallet/wallet.c]: nit: This and the next change could be added to the commit that added this fix in the first place.

File: wallet/wallet.c [hunk 2]
```diff
@@ -1378,19 +1385,20 @@ static void fixup_hin(struct wallet *wallet, struct htlc_in *hin)
...
-		   " subsituting temporary channel failure",
+		   " subsituting temporary node failure",
 		   hin->key.id, htlc_state_name(hin->hstate),
 		   hin->msatoshi,
 		   type_to_string(tmpctx, struct pubkey,
 				  &hin->key.channel->peer->id));
-}
 #endif
+}
+
```
Comment (cdecker) [inline wallet/wallet.c]: nit: extra newline

File: lightningd/htlc_end.h [hunk 3]
```diff
@@ -73,6 +73,9 @@ struct htlc_out {
 	/* If we fulfilled, here's the preimage. */
 	struct preimage *preimage;
 
+	/* Is this local?  Implies ->in is NULL. */
+	bool local;
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply lightningd/htlc_end.h]: Ack.

File: lightningd/peer_htlcs.c [hunk 4]
```diff
@@ -1716,6 +1716,13 @@ void htlcs_reconnect(struct lightningd *ld,
 				break;
 			}
 		}
+
+		if (!hout->in && !hout->preimage) {
+			/* FIXME: COMPAT_V061 only */
+			log_broken(ld->log,
+				   "Missing preimage for orphaned HTLC; replacing with zeros");
+			hout->preimage = talz(hout, struct preimage);
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply lightningd/peer_htlcs.c]: I'll add a FIXME.

File: tests/test_connection.py [hunk 5]
```diff
@@ -1374,3 +1374,64 @@ def test_fulfill_incoming_first(node_factory, bitcoind):
...
+                       'channel': inchans[i]},
+                      {'msatoshi': 100000000,
+                       'id': outnodes[i].info['id'],
+                       'delay': 5,
+                       'channel': outchans[i]}])
+        payment_hash.append(outnodes[i].rpc.invoice(100000000, "invoice2", "invoice2")['payment_hash'])
+
+    # sendpay is async.
+    for i in range(len(payment_hash)):
+        innodes[i // 2].rpc.sendpay(route[i], payment_hash[i])
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply tests/test_connection.py]: There are two payments per innode...

File: wallet/wallet.c (hunk 1 above)
Comment (rustyrussell) [reply wallet/wallet.c]: Yes, agreed.  I only realized it when I found that issue though...

File: wallet/wallet.c (hunk 2 above)
Comment (rustyrussell) [reply wallet/wallet.c]: I'll kill that, too.

**Comment by cdecker** [comment]
Awesome, thanks for the fixes @rustyrussell 

ACK https://github.com/ElementsProject/lightning/pull/2000/commits/14e27984ff457494cb2f9818df781e52c9fec9a5

The merge request should disappear when doing a `Rebase and merge`, hopefully :crossed_fingers:

File: tests/test_connection.py [hunk 6]
```diff
@@ -1374,3 +1374,64 @@ def test_fulfill_incoming_first(node_factory, bitcoind):
...
+    outnodes = node_factory.get_nodes(len(innodes), opts={'may_reconnect': True})
+    outchans = []
+    for n in outnodes:
+        n.rpc.connect(l1.info['id'], 'localhost', l1.port)
+        outchans.append(l1.fund_channel(n, 10**6))
+
+    # Manually create routes, get invoices
+    route = []
+    payment_hash = []
+    for i in range(len(innodes)):
```
[the remark this answers is not in the document]
Comment (niftynei) [reply tests/test_connection.py]: in the spirit of pythonicity, this is a bit nicer:

````
import itertools
for inode, outnode, outchan, inchan in itertools.izip(innodes,outnodes,outchans,inchans):
````

via https://stackoverflow.com/a/6498180

File: tests/test_connection.py [hunk 7]
```diff
@@ -1331,3 +1331,37 @@ def test_restart_multi_htlc_rexmit(node_factory, bitcoind, executor):
...
+@pytest.mark.xfail(strict=True)
+@unittest.skipIf(not DEVELOPER, "needs dev-disconnect")
+def test_fulfill_incoming_first(node_factory, bitcoind):
+    """Test that we handle the case where we completely resolve incoming htlc
+    before fulfilled outgoing htlc"""
+
+    # We agree on fee change first, then add HTLC, then remove; stop after remove.
+    disconnects = ['+WIRE_COMMITMENT_SIGNED*3']
+    # We manually reconnect l2 & l3, after 100 blocks; hence allowing manual
+    # reconnect, but disabling auto coneect, and massive cltv so 2/3 doesn't
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply tests/test_connection.py]: I had to rebase anyway due to CHANGELOG conflict, so I snuck this in too...  thanks!

**Pull request merged**